packages/devtools_app/lib/src/performance/performance_controller.dart

```diff
@@ -138,27 +168,84 @@ class PerformanceController
-      gcTimelineStream,
-    ]);
-    await toggleHttpRequestLogging(true);
-
-    // Initialize displayRefreshRate.
-    _displayRefreshRate.value =
-        await serviceManager.queryDisplayRefreshRate ?? defaultRefreshRate;
-    data?.displayRefreshRate = _displayRefreshRate.value;
+    if (!offlineMode) {
+      await serviceManager.onServiceAvailable;
+
+      data = serviceManager.connectedApp.isFlutterAppNow
+          ? PerformanceData(
+              displayRefreshRate: await serviceManager.queryDisplayRefreshRate,
+            )
+          : PerformanceData();
+
+      // Default to true for profile builds only.
+      _badgeTabForJankyFrames.value =
+          await serviceManager.connectedApp.isProfileBuild;
+
+      unawaited(allowedError(
+        serviceManager.service.setProfilePeriod(mediumProfilePeriod),
+        logError: false,
+      ));
+      await setTimelineStreams([
+        dartTimelineStream,
+        embedderTimelineStream,
+        gcTimelineStream,
+      ]);
+      await toggleHttpRequestLogging(true);
+
+      // Initialize displayRefreshRate.
+      _displayRefreshRate.value =
+          await serviceManager.queryDisplayRefreshRate ?? defaultRefreshRate;
+      data?.displayRefreshRate = _displayRefreshRate.value;
+
+      // Listen for Flutter.Frame events with frame timing data.
+      autoDispose(
+          serviceManager.service.onExtensionEventWithHistory.listen((event) {
+        if (event.extensionKind == 'Flutter.Frame') {
+          final frame = FlutterFrame.parse(event.extensionData.data);
+          addFrame(frame);
+        }
+      }));
+
+      // Load available timeline events.
+      final timeline = await serviceManager.service.getVMTimeline();
+      primeThreadIds(timeline);
+      for (final event in timeline.traceEvents) {
+        final eventWrapper = TraceEventWrapper(
+          TraceEvent(event.json),
+          DateTime.now().millisecondsSinceEpoch,
+        );
+        allTraceEvents.add(eventWrapper);
+      }
+
+      _processing.value = true;
+      await processTraceEvents(allTraceEvents);
+      _processing.value = false;
+
+      // Listen for new timeline events.
+      autoDispose(serviceManager.service.onTimelineEvent.listen((event) {
+        final eventBatch = <TraceEventWrapper>[];
+        if (event.json['kind'] == 'TimelineEvents') {
+          final List<dynamic> traceEvents = event.json['timelineEvents']
+              .map(
+                (e) => TraceEventWrapper(
+                  TraceEvent(e),
+                  DateTime.now().millisecondsSinceEpoch,
+                ),
+              )
+              .toList();
+          final List<TraceEventWrapper> wrappedTraceEvents =
+              traceEvents.cast<TraceEventWrapper>();
+          eventBatch.addAll(wrappedTraceEvents);
+        }
+        allTraceEvents.addAll(eventBatch);
+      }));
+    }
+  }
+
+  FutureOr<void> processAvailableEvents() async {
+    _processing.value = true;
```

Comment:
nit: add 
assert(!_processing.value) at the start of the method to make sure we don't get into a state where we have two pending processing requests.